podvm/qcow2/misc-settings.sh

```diff
@@ -36,8 +36,80 @@ then
     *)
         ;;
     esac
-# else if...
-#
+fi
+
+# Setup oneshot systemd service for AWS and Azure to enable NAT rules
+if [ "$CLOUD_PROVIDER" == "azure" ] || [ "$CLOUD_PROVIDER" == "aws" ]
+then
+# Create a oneshot systemd service to execute setup-nat-for-imds.sh
+# during first boot
+cat <<'END' > /etc/systemd/system/setup-nat-for-imds.service
+[Unit]
+Description=Setup NAT for IMDS
+After=kata-agent.service
+Wants=kata-agent.service
+
+[Service]
+Type=oneshot
+ExecStart=/usr/local/bin/setup-nat-for-imds.sh
+
+[Install]
+WantedBy=multi-user.target
+END
+
+# Add a link to the service in multi-user.target.wants
+ln -s /etc/systemd/system/setup-nat-for-imds.service /etc/systemd/system/multi-user.target.wants/setup-nat-for-imds.service
+
+# Create a script to setup NAT for IMDS
+cat <<'END' > /usr/local/bin/setup-nat-for-imds.sh
+#!/bin/bash
+# This script sets up NAT for IMDS on Azure/AWS
+# This is required for IMDS to work on Azure/AWS
+# This script is executed as a oneshot systemd service
+# during first boot
+
+set -euo pipefail
+
+IMDS_IP="169.254.169.254"
+DUMMY_IP="169.254.99.99"
+
+# trap errors
+trap 'echo "Error: $0:$LINENO stopped"; exit 1' ERR INT
+
+# Function to setup veth pair
+function setup_proxy_arp() {
+  local pod_ip
+  pod_ip=$(ip netns exec podns ip route get "$IMDS_IP" | awk 'NR == 1 { print $7 }')
```

Comment:
It fails here when when it run by systemd, i think the reason is network isn't ready after kata-agent is up but only after some other grpc call arrives (createContainer?)
FWIW what did work for was to get the ip from daemon.json (something like `pod_ip=$(grep podip /peerpod/daemon.json | grep -o -E '([0-9]{1,3}[\.]){3}[0-9]{1,3}')`)
and making sure the oneshot is running after the cloud-init.service

Reply:
I think we can fix the issue by setting dependency to `agent-protocol-forwarder.service` instead of `kata-agent.service`.

`agent-protocol-forwarder.service` is launched after `kata-agent.service`, and notifies `systemd` when it completes network initialization of the `podns` network namespace.

https://github.com/confidential-containers/cloud-api-adaptor/blob/a771f96274c8b98f9f450b9477ad0572902ae94e/cmd/starter.go#L96-L115

The `podns` network namespace is initialized here.
https://github.com/confidential-containers/cloud-api-adaptor/blob/a771f96274c8b98f9f450b9477ad0572902ae94e/pkg/forwarder/forwarder.go#L91

And, `systemd` is notified here.
https://github.com/confidential-containers/cloud-api-adaptor/blob/a771f96274c8b98f9f450b9477ad0572902ae94e/pkg/forwarder/forwarder.go#L157

Reply:
We may also need to specify the following setting as discussed in https://github.com/confidential-containers/cloud-api-adaptor/pull/389

```
DefaultDependencies=no
```

Reply:
This works now.